Tests/MessagingPush/PushHandling/PushEventHandlerTest.swift

```diff
@@ -0,0 +1,41 @@
+@testable import CioMessagingPush
+import Foundation
+import SharedTests
+import XCTest
+
+class PushEventHandlerTest: UnitTest {
+    private var pushEventHandler: PushEventHandler!
+
+    private var pushEventHandlerProxy: PushEventHandlerProxy {
+        diGraph.pushEventHandlerProxy
+    }
+
+    private var pushClickHandler = PushClickHandlerMock()
+
+    override func setUp() {
+        super.setUp()
+
+        pushEventHandler = IOSPushEventListener(jsonAdapter: diGraph.jsonAdapter, pushEventHandlerProxy: pushEventHandlerProxy, moduleConfig: MessagingPushConfigOptions(), pushClickHandler: pushClickHandler, pushHistory: diGraph.pushHistory, logger: log)
+    }
+
+    // MARK: onPushAction
+
+    /*
+     The CIO SDK push event handler forwards push notifications to other push event handlers in the host app.
+
+     But, what if the CIO SDK event handler forwards a push event back to the CIO SDK event handler? This could cause an infinite loop.
+
+     This test simulates that scenario to make sure that an inifinite loop would not happen.
+     */
+    func test_onPushAction_expectNoInfiniteLoopIfSdkForwardsPushEventBacktoSdkAgain() {
+        // The push event handler proxy is what forwards push events to other push event handlers in the app.
+        // To test if an infinite loop would happen, add ourself as a push event handler to get events forwarded to.
+        pushEventHandlerProxy.addPushEventHandler(pushEventHandler)
+
+        // Make sure push is not from CIO otherwise event does not get forwarded.
+        let givenPush = PushNotificationStub.getPushNotSentFromCIO()
+
+        // Send the CIO SDK push event handler an event. If an infinite loop occurs, this test will timeout or crash.
```

Comment:
Is it possible to mock push history and induce a timeout/crash to more effectively validate our assumptions?

Reply:
Same, should we timeout it something like this? 

```
let expectation = expectation(description: "Completion handler should be called")

    // ... Set up your asynchronous operation ...

    // This is crucial for timeout assertions
    waitForExpectations(timeout: 5.0) { error in
        if let error = error {
            XCTFail("Timeout error: \(error)")
        } 
    }
```

Reply:
I was wondering the same thing when I wrote this test. But, I currently cannot think of any way to do this. None of the code under test is async and I don't see how mocking the push history will help, especially when the push history is the reason why an infinite loop does not exist. Therefore, I think it's important to run the real implementation of push history. 

The test function asserts that an infinite loop will not occur. If an infinite loop does occur, then a stackoverflow will occur and cause the test to crash. I think that is enough to validate the test?

Reply:
Since the code that prevents the infinite loop is:

```swift
guard !pushHistory.hasHandledPush(pushEvent: .didReceive, pushId: push.pushId, pushDeliveryDate: dateWhenPushDelivered)
```

And considering `PushHistory` is our class too, if `pushHistory.hasHandledPush` keeps returning `false`, shouldn't that trigger timeout/crash validating our assumptions about this behavior?

Reply:
If we wanted to assert that the push history feature is specifically the reason why an infinite loop does not happen, then yes, I can see how that idea could work. 

However, I don't think that is valuable to the test suite. We do not care how the infinite loop is prevented, we only care that an infinite loop does not occur. 

Same goes for [this test](https://github.com/customerio/customerio-ios/blob/216341c932d29caf334da53df3b4ab37b97bc94a/Tests/MessagingPush/PushHandling/AutomaticPushClickedIntegrationTest.swift#L88-L121) that is also testing an infinite loop does not occur. The test sets up a scenario that could create an infinite loop. If this test ever fails, then we have to fix the code and prevent the infinite loop from happening.